[new] sqldeveloper derivation

ardumont opened the pull request:
I did not yet successfully found a way to use the `sqlplus` installed by the derivation oracle-xe.
As I do not master oracle's setup intricacies (nor do I really want to), I settle on this.

I suppose I'm not the only one who will someday need it.

ardumont committed:
Add sqldeveloper derivation

ardumont committed:
Fix frei0r url

ardumont commented:
Rebased to latest code to solve conflicts.

ardumont commented on pkgs/development/libraries/frei0r/default.nix:
```diff
@@ -5,7 +5,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "https://files.dyne.org/frei0r/releases/frei0r-plugins-1.4.tar.gz";
+    url = "https://files.dyne.org/frei0r/releases/${name}.tar.gz";
```
I initially had to fix this url which had changed.
I also avoided repetition by using the `name`.
After rebase, I solve the conflict letting only the factorisation appear.
I can remove this commit if you see this as inconvenient commit.

ardumont commented:
After needing some time for building on latest commit, it still works.

![screenshot_2015-03-13_07-24-50](https://cloud.githubusercontent.com/assets/718812/6634025/69a5b8ea-c952-11e4-877b-1bce2f6fe70f.png)

ardumont commented on pkgs/development/tools/database/sqldeveloper/default.nix:
```diff
@@ -0,0 +1,48 @@
+    chmod +x $out/sqldeveloper/bin/sqldeveloper
+
+    wrapProgram $out/bin/sqldeveloper \
+      --set JAVA_HOME "${oraclejdk7}"
```
Starting this sqldeveloper version, it explicits the use of an oracle version superior or equal to 7 but not 8... So 7 it is :D 

Thanks to nixos, this is so simple to do!

lethalman replied:
First of all why the jdk instead of jre? Also, I'd like to add a jdk7 attribute to all-packages.nix, so that one can use either openjdk or oraclejdk.

ardumont replied:
> First of all why the jdk instead of jre

Good question. JRE should indeed suffice.

ardumont replied:
I remember now, it's because, it's sqldeveloper which claim to be in need of a jdk

Here is an extract of a configuration file (I do not use here because we set the JAVA_HOME) which speaks of JDK:

```txt
#
# By default, the product launcher will search for a JDK to use, and if none
# can be found, it will ask for the location of a JDK and store its location
# in this file. If a particular JDK should be used instead, uncomment the
# line below and set the path to your preferred JDK.
#
```

ardumont replied:
Also, here is the message when installed without the JAVA_HOME set is this:

``` sh
# tony at lenovo in ~/repo/perso/nixpkgs on git:add-sqldeveloper-derivation x [12:20:08]
$ ./result/bin/sqldeveloper

 Oracle SQL Developer
 Copyright (c) 1997, 2014, Oracle and/or its affiliates. All rights reserved.

Type the full pathname of a JDK installation (or Ctrl-C to quit), the path will be stored in /home/tony/.sqldeveloper/4.0.0/product.conf
```

So it asks for a jdk.
But after modification, the jre seems to work fine (at least for the startup).

ardumont replied:
WAIT.

Nope. I tested with the oraclejre7 and it does not work!
It asks for a jdk.

Tested again by removing the file `~/.sqldeveloper/4.0.0/product.conf` first:
- with oraclejre7: FAIL, it asks the first time (i want to avoid this)
- with oraclejdk7: Ok.

ardumont commented on pkgs/development/tools/database/sqldeveloper/default.nix:
```diff
@@ -0,0 +1,48 @@
+  buildCommand = ''
+    mkdir -p $out/bin
+    # patch to be able to install a sqldeveloper wrapper script compliant with nix's bin folder once installed
+    echo -e '#!/bin/bash\ncd "`dirname $0`"/../sqldeveloper/bin && bash sqldeveloper $*' >> $out/bin/sqldeveloper
```
If I want to see sqldeveloper installed in PATH once installed in nix, it must be in a bin/ folder.
If I simply move the sqldeveloper.sh provided, it won't work because it hardcoded value. This is not free software so no contribution.

So I create an equivalent one with the .. path.
Also as a convenience, I removed the suffix `.sh`.

This `sqldeveloper` will be the one wrapped line 28.

lethalman replied:
Do not use /bin/bash please, use `${bash}/bin/bash`.

ardumont replied:
> Do not use /bin/bash please, use ${bash}/bin/bash

Thanks.

> also why do you need that directory check?

I do not understand the question.

lethalman replied:
The `dirname $0`"/../sqldeveloper/bin

ardumont replied:
It's not a check, it's to make it ok between nix (to be able to call sqldeveloper without it's absolute path) and the slightly modified `sqldeveloper.sh` wrapper (which do not permit extensibility with the wrapProgram).

From `$out/bin`, the `dirname $0"/../sqldeveloper/bin` folder is the actual sqldeveloper binary.

I do not want to differ too much from the current nixos approach.
But I do in generating this bash script.
But because the existing one cannot be inside `$out/bin` or else it won't work.

I'm not happy to generate it myself the `sqldeveloper.sh` so I try to make it as much as possible the same as the original one (to ease the maintenance later).

lethalman commented:
So is this ready to merge?

ardumont commented:
> So is this ready to merge?

@lethalman for me yes. I just describe my reasoning to give you insight on some choices I made.

ardumont committed:
Adapt code according to @lethalman's review

- Use ${bash}/bin/bash
- We cannot use oraclejre7 because sqldeveloper wants a jdk.

lethalman commented:
No worries, just use jdk :)

ardumont commented:
To be clear, the jre is just ignored and the first time you start sqldeveloper, you will be prompt to enter your jdk path...

ardumont commented:
I also tested with `jdk` (icedtea) and it's ignored.
It really wants an `oraclejdk`.

ardumont commented:
Do you want me to squash the last 2 commits?

lethalman commented:
No it's ok. Final note, you have still a bash call without the nix store path there.

ardumont commented:
> Final note, you have still a bash call without the nix store path there.

Right!

ardumont commented:
I amended the last commit!

Thanks for the review (and your blog posts :beers:)

lethalman commented:
Ok thanks.

bluescreen303 commented:
@ardumont it seems version 4.1 is out. Luckily the oracle site still has a "previous release" link to the version you packaged, but that will go with the next release. Would be great if you could do an update :)

ardumont commented:
Hello,

@bluescreen303 I'll see what I can do :D 

Cheers,
